src/core/batch.rkt

```diff
@@ -72,163 +76,169 @@
-    (define final-idx
-      (let loop ([expr replacement])
-        (match expr
-          [(batchref b* idx)
-           (unless (eq? b* b)
-             (error 'batch-replace "Replacement ~a references the wrong batch ~a" replacement b*))
-           (when (= -1 (vector-ref mapping idx))
-             (error 'batch-replace "Replacement ~a references unknown index ~a" replacement idx))
-           (vector-ref mapping idx)]
-          [_ (batch-push! out (expr-recurse expr loop))])))
-    (vector-set! mapping idx final-idx))
-  (define roots (vector-map (curry vector-ref mapping) (batch-roots b)))
-  (set-batch-roots! out roots)
-  out)
-
-;; Function returns indices of alive nodes within a batch for given roots,
-;;   where alive node is a child of a root + meets a condition - (condition node)
-(define (batch-alive-nodes batch
-                           [roots (batch-roots batch)]
-                           #:keep-vars-alive [keep-vars-alive #f]
-                           #:condition [condition (const #t)])
+  (define brfs
+    (for/list ([expr (in-list exprs)])
+      (batch-add! out expr)))
+  (values out brfs))
+
+(define (batch->progs b brfs)
+  (brfs-belong-to-batch? b brfs)
+  (map (batch-exprs b) brfs))
+
+;; batch-map does not iterate over nodes that are not a child of brf
+;; A lot of parts of Herbie rely on that
+(define (batch-map batch f)
   (define len (batch-length batch))
-  (define alive-mask (make-vector len #f))
-  (for ([root (in-vector roots)])
-    (vector-set! alive-mask root #t))
-  (for ([i (in-range (- len 1) -1 -1)]
-        [node (in-batch batch (- len 1) -1 -1)]
-        [alv (in-vector alive-mask (- len 1) -1 -1)]
-        #:when (or (and alv (condition node)) (and keep-vars-alive (symbol? node))))
-    (unless alv ; if keep-vars-alive then alv may not be #t, making sure it's #t
-      (vector-set! alive-mask i #t))
-    (expr-recurse node
-                  (λ (n)
-                    (when (condition (batch-ref batch n))
-                      (vector-set! alive-mask n #t)))))
-  ; Return indices of alive nodes in ascending order
-  (for/vector ([alv (in-vector alive-mask)]
-               [i (in-naturals)]
-               #:when alv)
-    i))
+  (define out (make-dvector))
+  (define visited (make-dvector))
+  (λ (brf)
+    (match-define (batchref b idx) brf)
+    (cond
+      ; Little check
+      [(or (not (equal? b batch)) (>= idx len)) (error 'batch-map "Inappropriate batchref is passed")]
+      [else
+       (let loop ([idx idx])
+         (match (and (> (dvector-capacity visited) idx) (dvector-ref visited idx))
+           [#t (dvector-ref out idx)]
+           [_
+            (define node (batch-ref batch idx))
+            (define res (f (λ (x) (loop x)) node))
+            (dvector-set! out idx res)
+            (dvector-set! visited idx #t)
+            res]))])))
+
+(define (batch-apply b brfs f)
+  (define out (batch-empty))
+  (define apply-f
+    (batch-map b
+               (λ (remap node)
```

Comment:
Purely stylistic, but in general instead of doing:
```
(define some-f
  (batch-map b
             (lambda (recur args ...)
               ...)))
... (map some-f brfs) ...
```
it would look cleaner to write:
```
(define (some-f recur args ...)
  ...)
... (map (batch-map b some-f) brfs) ...
```